generator/generator.go

```diff
@@ -49,3 +56,119 @@ func (g *Generator) writeFile(pkg *Package) (err error) {
 }
+
+// Timestamper is a function that returns the current time.
+type Timestamper func() time.Time
+
+// MigrationGenerator is a generator of migrations.
+type MigrationGenerator struct {
+	name string
+	dir  string
+	now  Timestamper
+}
+
+type migrationFileType string
+
+const (
+	migrationUp   = migrationFileType("up.sql")
+	migrationDown = migrationFileType("down.sql")
+	migrationLock = migrationFileType("lock.json")
+)
+
+// NewMigrationGenerator returns a new migration generator with the given
+// migrations directory.
+func NewMigrationGenerator(name, dir string) *MigrationGenerator {
+	return &MigrationGenerator{slugify(name), dir, time.Now}
+}
+
+// Build creates a new migration from a set of scanned packages.
+func (g *MigrationGenerator) Build(pkgs ...*Package) (*Migration, error) {
+	old, err := g.LoadLock()
+	if err != nil {
+		return nil, err
+	}
+
+	new, err := SchemaFromPackages(pkgs...)
+	if err != nil {
+		return nil, err
+	}
+
+	return NewMigration(old, new), nil
+}
+
+// Generate will generate the given migration.
+func (g *MigrationGenerator) Generate(migration *Migration) error {
+	return g.writeMigration(migration)
+}
+
+// LoadLock loads the lock file.
+func (g *MigrationGenerator) LoadLock() (*ModelSchema, error) {
+	bytes, err := ioutil.ReadFile(filepath.Join(g.dir, string(migrationLock)))
+	if os.IsNotExist(err) {
+		return new(ModelSchema), nil
+	} else if err != nil {
+		return nil, fmt.Errorf("error opening lock file: %s", err)
+	}
+
+	var schema ModelSchema
+	if err := json.Unmarshal(bytes, &schema); err != nil {
+		return nil, fmt.Errorf("error unmarshaling lock schema: %s", err)
+	}
+
+	return &schema, nil
+}
+
+func (g *MigrationGenerator) writeMigration(migration *Migration) error {
+	t := g.now()
+	files := []struct {
+		file    string
+		content encoding.TextMarshaler
+	}{
+		{filepath.Join(g.dir, string(migrationLock)), migration.Lock},
+		{g.migrationFile(migrationDown, t), migration.Down},
+		{g.migrationFile(migrationUp, t), migration.Up},
+	}
+
+	for _, f := range files {
+		if err := g.createFile(f.file, f.content); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (g *MigrationGenerator) migrationFile(typ migrationFileType, t time.Time) string {
+	return filepath.Join(g.dir, fmt.Sprintf("%d_%s.%s", t.Unix(), g.name, typ))
+}
+
+func (g *MigrationGenerator) createFile(filename string, marshaler encoding.TextMarshaler) error {
+	data, err := marshaler.MarshalText()
+	if err != nil {
+		return err
+	}
+
+	f, err := os.OpenFile(filename, os.O_TRUNC|os.O_CREATE|os.O_WRONLY, 0755)
+	if err != nil {
+		return fmt.Errorf("error opening file: %s: %s", filename, err)
+	}
+
+	defer f.Close()
+	if _, err := f.Write(data); err != nil {
+		return fmt.Errorf("error writing file: %s: %s", filename, err)
+	}
+
+	return nil
+}
+
+func slugify(str string) string {
+	var buf bytes.Buffer
+	for _, r := range strings.ToLower(str) {
+		if ('a' <= r && r <= 'z') || ('0' <= r && r <= '9') {
+			buf.WriteRune(r)
```

Comment:
But it's always nil https://golang.org/src/bytes/buffer.go?s=8208:8261#L237

All the writes to `bytes.Buffer` return nil